src/tests/test_types.cpp

```diff
@@ -264,4 +264,109 @@ INSTANTIATE_TEST_SUITE_P(TestStringSuite, TestString,
+
+DatePtr DatePtrFromString(std::string const& s, std::unique_ptr<mo::DateType> const& date_type) {
+    return {date_type->MakeValue(boost::gregorian::from_simple_string(s)),
+            date_type->GetDeleater()};
+}
+
+TEST(TestDateSuite, Compare) {
+    std::unique_ptr<mo::DateType> date_type(
+            mo::CreateSpecificType<mo::DateType>(mo::TypeId::kDate, true));
+    auto test = [date_type = std::move(date_type)](
+                        std::string const& date1, std::string const& date2, mo::CompareResult res) {
+        auto ptr1 = DatePtrFromString(date1, date_type);
+        auto ptr2 = DatePtrFromString(date2, date_type);
+        EXPECT_EQ(date_type->Compare(ptr1.get(), ptr2.get()), res);
+    };
+    test("2004-07-28", "2004-07-28", mo::CompareResult::kEqual);
+    test("2005-01-20", "2004-07-28", mo::CompareResult::kGreater);
+    test("2023-09-01", "2024-10-25", mo::CompareResult::kLess);
+    test("2023-10-28", "2023-10-28", mo::CompareResult::kEqual);
+    test("2203-09-01", "2024-10-25", mo::CompareResult::kGreater);
+    test("1991-11-20", "2200-07-28", mo::CompareResult::kLess);
+}
+
+TEST(TestDateSuite, Dist) {
+    std::unique_ptr<mo::DateType> date_type(
+            mo::CreateSpecificType<mo::DateType>(mo::TypeId::kDate, true));
+    auto test = [date_type = std::move(date_type)](std::string const& date1,
+                                                   std::string const& date2, long res) {
+        auto left_val = DatePtrFromString(date1, date_type);
+        auto right_val = DatePtrFromString(date2, date_type);
+        EXPECT_EQ(long(date_type->Dist(left_val.get(), right_val.get())), res);
+    };
+    test("2004-07-28", "2004-07-28", 0);
+    test("2005-01-20", "2004-07-28", 176);
+    test("1991-11-20", "2200-07-28", 76221);
+    test("1941-06-22", "1945-05-09", 1417);
+    test("2023-10-28", "2297-08-12", 100000);
+}
+struct TestDateArithmeticsParam {
+    std::string const date1;
+    std::string const date2;
+    long const dist;
+    TestDateArithmeticsParam(std::string l, std::string r, long dist)
+        : date1(std::move(l)), date2(std::move(r)), dist(dist){};
+};
+
+class TestDateArithmetics : public ::testing::TestWithParam<TestDateArithmeticsParam> {};
+TEST_P(TestDateArithmetics, SubtractDelta) {
+    std::unique_ptr<mo::DateType> date_type(
+            mo::CreateSpecificType<mo::DateType>(mo::TypeId::kDate, true));
+
+    auto const& [date1, date2, delta_] = GetParam();
+    mo::DateType::Delta delta(delta_);
+
+    auto initial_date = DatePtrFromString(date1, date_type);
+    auto estimated_result = DatePtrFromString(date2, date_type);
+    auto actual_result =
+            DatePtr(date_type->SubDelta(initial_date.get(), delta), date_type->GetDeleater());
+
+    EXPECT_EQ(date_type->Compare(estimated_result.get(), actual_result.get()),
+              mo::CompareResult::kEqual);
+}
+
+TEST_P(TestDateArithmetics, AddDelta) {
```

Comment:
This test case has almost the same code as the test above. Common logic should be moved to a separate function which expects tested method, dates and probably an initial date as params.